### DISCUSSION TITLE
top-down model of five fish maintaining an extremely high error rate

### DISCUSSION THREAD

**Discussion by mollywestbrook** [talmolab/sleap#1392] [Help!]
Hi all, was hoping to get some pointers as to why our model is doing poorly without improvement. A screengrab of our set-up and skeleton: 
![image](https://github.com/talmolab/sleap/assets/48096260/69e0e88e-33ad-4997-86bf-736c601d8cbe)

These are the most recent error outputs of the latest model, as well as the hyperparameters listed below:
![image](https://github.com/talmolab/sleap/assets/48096260/c7f9d281-b58c-41d3-9453-c8a8026356b6)
![image](https://github.com/talmolab/sleap/assets/48096260/eb5bf7cf-462d-4861-b3af-59e366a82ce9)
![image](https://github.com/talmolab/sleap/assets/48096260/7727fb0e-9f00-41dd-bc64-5f891fdd6c35)
![image](https://github.com/talmolab/sleap/assets/48096260/6e1084f8-eeba-4efb-8e75-129b8cb1fac7)

We also tried simplifying the skeleton, with little success (same hyperparameters, only change was the skeleton).
![image](https://github.com/talmolab/sleap/assets/48096260/1d9eb5a0-8a60-4a0e-ab1d-fb3ec60cf929)
![image](https://github.com/talmolab/sleap/assets/48096260/d1e5d1d1-df54-4dba-b57c-44bc7b70d1a1)

We have almost 700 frames labelled across 13 videos. The model does fine when we output 20 random frames, the predictions seem to be doing quite well. Likewise, when we track some sample videos, the predicted instances of the skeleton tend to be quite accurate, only the tracker will also tend to track parts of the tank environment, and has a lot of trouble maintaining track identities over time. In addition, the error parameters tend to be rather concerning, with high distances and extremely low precision/recall scores. Not sure whether or not this is a big deal? I tend to see upwards of 100 identities, and a lot of the identities will be on the environment, rather than on the animal:
![image](https://github.com/talmolab/sleap/assets/48096260/4247aa33-dbf9-4dbb-ab1c-39d5ac1e759a)

We've been doing some evaluations of model versions on a standardized video, to demonstrate: 
![image](https://github.com/talmolab/sleap/assets/48096260/89137717-e14f-41ef-ae28-380db4d28815)
and a breakdown of how nodes are doing across videos, taken from the .h5 scores:
![image](https://github.com/talmolab/sleap/assets/48096260/dce216d6-52c5-4b79-9f8e-145af7604cea)

Was curious if adjusting the hyperparameters as suggested in [1369](https://github.com/talmolab/sleap/discussions/1369) or in [1340 ](https://github.com/talmolab/sleap/discussions/1340) might do the trick, or if we should just add another large chunk of frames to the training dataset for improvement. Thanks for any help you could provide!

**Comment by talmo** [comment]
Hi @mollywestbrook,

Thanks for providing the detailed results on your experiments so far!! Super helpful for diagnostics.

It's definitely strange to see such a wide distribution of errors when the qualitative results look so good. We usually see this when there are a handful of frames that are labeled incorrectly or inconsistently, for example, when not all animals in a frame were labeled, or when nodes were marked as "not visible" when they should have been "visible". In both of these cases, parts of the image that should have been predicted on are instead being treated as negative examples, which can really confuse the model. Any chance this could be the case?

You could do a quick inspection of the labeled frames to see if there's any that might be missing labels. Alternatively, as a sanity check you could start a new project where you label ~10-20 frames really carefully and consistently and see what the error distribution looks like. If it's less variable than the bigger project despite having fewer frames, it's a good indicator that there might be some outlier labels in the original dataset.

It's also especially strange that the "head" node has such a high error rate. If this is being used as the anchor node for top-down models, it should have near 0 error rate by virtue of how the models are constructed (the anchor node is always at the center of the crop that's used by the centered instance model so it should be _really_ easy to find). Do you mind exporting a training package with embedded images (Predict --> Export training package) and [sharing it with us via this form](https://forms.gle/B7d5M4j5mwC9n2qj7)? I suspect there may be a bug related to the skeleton editing you did as part of your experimentation.

While we have a look at that, some model parameter tweaks I'd recommend:
- Set rotation augmentation to -180 to 180
- Set output stride to 2 in the centered instance model

Let us know when you get a chance to try out some of these troubleshooting steps and we'll go from there.

Cheers,

Talmo

**Reply by talmo** [reply]
Oh and I forgot to touch on identities:

You can now set the maximum number of identities right in the Run Prediction window if you update to SLEAP v1.3.1 -- this will help cut down the number of tracks (though you'll still need to proofread for potential identity switches).

The total number of tracks might still be greater than the number you specified, but it'll help, especially with background detections.

**Reply by mollywestbrook** [reply]
Talmo,

Thank you as always for your super thorough and helpful response! We have a training package we just exported for remote training (this is without the hyperparameter changes you suggest, but we'll plan to implement those soon), so I will submit through the form you sent. In the meantime, we'll start a fresh project for comparison.

In some of the videos, fish tails are not visible due to camera exposure, so in these instances I've labelled them as 'not visible'. From your suggestion, it seems like these points should be labelled as visible to our best guess anyway? Depending on the results of the fresh project test, I can go back and label nodes as visible in our existing project. As for setting the maximum number of identities, at this stage I have found it difficult to proofread when we have implemented a cap, so until we fix the identity problem I will leave it uncapped.  

I will update the thread once we have some answers. Again, thank you!

**Reply by mollywestbrook** [reply]
Hi Talmo,

We tried a couple of your suggested implements. First, went through and created a fresh new project, labelled 20 frames, and had the model label some frame predictions. The scores tended to vary heavily and it was unable to label a couple of the instances. Image below. 
![framescores](https://github.com/talmolab/sleap/assets/48096260/dbb30723-601d-45c9-adcf-b4df99129724)

I also checked the error metrics for the new 20-frame model and there wasn't much improvement:
Error distance (50%): 2.214885882674386
Error distance (90%): 11.474636767450617
Error distance (95%): 13.902026145472966
mAP: 0.057301980198019804
mAR: 0.11000000000000001

Compared to our 90-frame model, fairly early on in the process:
Error distance (50%): 2.056
Error distance (90%): 5.26
Error distance (95%): 7.28
mAP: 0.211
mAR: 0.329

Then went through our original ~700 frame model and corrected all instances that were mislabelled or had hidden nodes (there actually ended up being several of these, so hopefully this will make a difference). I will update the thread with those results as soon as our new models are finished training. Still have not implemented hyperparameter changes yet, but we'll try that next.

**Reply by mollywestbrook** [reply]
Another update. The newly trained models from the corrected project finished training, and the error rate improvement was negligible. The frame scores of 20 random frames were still quite good:
![frame scores 072523](https://github.com/talmolab/sleap/assets/48096260/7146ce3c-eda8-47de-9500-0f8dcdc82741)

so the model is good at predicting single frames, but the error scores are still very, very low. Metrics:
![image](https://github.com/talmolab/sleap/assets/48096260/1e969009-eb7f-4c0f-b475-d9ca9705441a)
![image](https://github.com/talmolab/sleap/assets/48096260/bf36c0f0-714f-4206-b231-0d56429bb634)


Still can't figure this one out. Any ideas? We retrained the exact same labels set with the suggested hyperparameters. I'll post their metrics tomorrow.

**Reply by mollywestbrook** [reply]
Reran our original ~700 frame model with the updated hyperparameters, in addition to correcting any mislabelled frames. It does seem like our centroid model is being thrown by a small percentage of extreme outliers, but this contradicts the corrections we did. Our centered instance model node distance is definitely dropping. At this point will more frames make a difference? We'll also try and track some videos with our latest model and see how they do. 
![image](https://github.com/talmolab/sleap/assets/48096260/8e0f5f07-1732-4a77-a5b4-76481325e632)
![image](https://github.com/talmolab/sleap/assets/48096260/fd171dbf-01b3-44eb-85ed-23249a3e4251)
![updated centered instance](https://github.com/talmolab/sleap/assets/48096260/a06a7522-9eef-42be-b2ab-3a13ea999209)
![updated centered instance errors](https://github.com/talmolab/sleap/assets/48096260/276cfd54-e07a-4179-8f4e-827f1f6895d0)
